### PR TITLE
Async Close Fixes (Issues #49 & #72)

### PR DESCRIPTION
**Pull request by chrisdennis**

### CODE REVIEW

File: src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java [hunk 1]
```diff
@@ -526,62 +480,36 @@ public void bind(Owner m) {
...
-                + "Consider increasing the -XX:MaxDirectMemorySize property to "
-                + "allow enough space for the FileChannel transfer buffers");
-        throw e;
+      long position = filePosition + address;
+      FileWriteTask pending = pendingWrites.get(position);
+      if (pending == null) {
+        ByteBuffer lengths = readBuffer(position + KEY_LENGTH_OFFSET, 8);
+        int keyLength = lengths.getInt(0);
+        int valueLength = lengths.getInt(VALUE_LENGTH_OFFSET - KEY_LENGTH_OFFSET);
+        return readBuffer(position + keyLength + KEY_DATA_OFFSET, valueLength);
```
Comment (cljohnso) [inline src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java:510]: This unnerves me a bit ... the original code read two ints each from a location based on position.  This code reads two ints with the first based on the original position and the second based not on position but on position of the first field -- an assumption is made about the relationship between the two fields that is not consistent with the original code -- `lengths` only contains 8 bytes of data ... what happens if the key and value length fields are not adjacent?

File: src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java [hunk 2]
```diff
@@ -624,41 +552,34 @@ Long writeMappingBuffers(ByteBuffer keyBuffer, ByteBuffer valueBuffer, int pojoH
...
-          int keyLength = readIntFromChannel(position + KEY_LENGTH_OFFSET);
-          int valueLength = readIntFromChannel(position + VALUE_LENGTH_OFFSET);
-          return getDiskWriteContext(position + keyLength + KEY_DATA_OFFSET, valueLength);
+      long position = filePosition + address;
+      FileWriteTask pending = pendingWrites.get(position);
+      if (pending == null) {
+        ByteBuffer lengths = readBuffer(position + KEY_LENGTH_OFFSET, 8);
+        int keyLength = lengths.getInt(0);
+        int valueLength = lengths.getInt(VALUE_LENGTH_OFFSET - KEY_LENGTH_OFFSET);
+        return getDiskWriteContext(position + keyLength + KEY_DATA_OFFSET, valueLength);
```
Comment (cljohnso) [inline src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java:576]: Same as the comment on `readValueBuffer` above.

File: src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java [hunk 3]
```diff
@@ -701,22 +622,19 @@ public void flush() {
...
-        throw new RuntimeException(e);
+      long position = filePosition + address;
+      int keyLength;
+      int valueLength;
+      FileWriteTask pending = pendingWrites.remove(position);
+      if (pending == null) {
+        ByteBuffer lengths = readBuffer(position + KEY_LENGTH_OFFSET, 8);
+        keyLength = lengths.getInt(0);
+        valueLength = lengths.getInt(VALUE_LENGTH_OFFSET - KEY_LENGTH_OFFSET);
+      } else {
```
Comment (cljohnso) [inline src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java:633]: Ditto ...

**Review by cljohnso — CHANGES_REQUESTED** [review]

File: src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java (hunk 1 above)
Comment (chrisdennis) [reply src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java:510]: Never trust git... this is a rebasing induced mistake. I'll do the rebase more carefully and it should clean up.

File: src/main/java/org/terracotta/offheapstore/util/ReopeningIoDevice.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import java.io.IOException;
+import java.nio.channels.ClosedChannelException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+public final class ReopeningIoDevice<T extends Closeable> implements Closeable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReopeningIoDevice.class);
+
+  public static <T extends Closeable> ReopeningIoDevice<T> createIoDevice(Supplier<T> deviceFactory) {
```
[the remark this answers is not in the document]
Comment (cljohnso) [reply src/main/java/org/terracotta/offheapstore/util/ReopeningIoDevice.java:31]: Or even `InterruptibleChannel`?

File: src/main/java/org/terracotta/offheapstore/util/ReopeningIoDevice.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        } else {
+          try {
+            return operation.apply(current);
+          } catch (ClosedChannelException e) {
+            interrupted |= Thread.interrupted();
+            T newDevice = deviceFactory.get();
+            if (!deviceReference.compareAndSet(current, newDevice)) {
+              newDevice.close();
+            } else {
+              LOGGER.info("Replacing IO device " + current + " with " + newDevice + " due to closure (likely due to interrupt)");
```
[the remark this answers is not in the document]
Comment (chrisdennis) [reply src/main/java/org/terracotta/offheapstore/util/ReopeningIoDevice.java]: Right... except we don't actually know why the replacement is happening. We might not be the trigger.

File: src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java (hunk 2 above)
Comment (chrisdennis) [reply src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java]: More rebasing woes.

File: src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java (hunk 3 above)
Comment (chrisdennis) [reply src/main/java/org/terracotta/offheapstore/disk/storage/FileBackedStorageEngine.java]: Rebasing ditto.

File: src/main/java/org/terracotta/offheapstore/util/ReopeningIoDevice.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cljohnso) [reply src/main/java/org/terracotta/offheapstore/util/ReopeningIoDevice.java]: Not saying that interrupt == trigger, just it's a bit of information that may come in handy when someone complains.  We know if the the thread was interrupted or not; just include that in the message.

**Review by cljohnso — APPROVED** [review]
Still have an open comment ... but it's not critical.

**Review by cljohnso — APPROVED** [review]

**Pull request merged**